Add documentation for Kubeflow Pipelines Kubernetes Native API mode

VaniHaripriya opened the pull request:
## ✏️ Summary of Changes

- Updated the README with instructions for deploying in the Kubernetes native API mode.
- Added instructions to compile pipelines to the Kubernetes manifest format with the KFP SDK.

## ✅ Contributor Checklist
  - [x] All commits are [_signed-off_](https://github.com/kubeflow/community/tree/master/dco-signoff-hook#signing-off-commits) to satisfy the DCO check.

gmfrasca commented on README.md:
```diff
@@ -412,6 +412,92 @@ kustomize build applications/pipeline/upstream/env/cert-manager/platform-agnosti
+**Key Differences in Kubernetes Native API Mode:**
+
+- Pipeline definitions are stored as `Pipeline` and `PipelineVersion` Custom Resources in Kubernetes
+- No external database required for pipeline storage (metadata still uses external storage)
```
either `metadata still stored in external database` or `mlmd still uses external storage` - just `metadata` sounds like the concept of metadata, not the ml-metadata component/dependency

juliusvonkohout replied:
Lets ue ml-metadata or whatever the official name is, no unnecessary abbreviations.

gmfrasca commented on README.md:
````diff
@@ -412,6 +412,92 @@ kustomize build applications/pipeline/upstream/env/cert-manager/platform-agnosti
+**Working with Pipelines in Kubernetes Native Mode:**
+
+```sh
+# After compilation, upload the pipeline through KFP API or UI
````
we should provide a summary in this section, so we could move these first few comments outside of the example commands instead

gmfrasca commented on README.md:
````diff
@@ -412,6 +412,92 @@ kustomize build applications/pipeline/upstream/env/cert-manager/platform-agnosti
+)
+```
+
+**Working with Pipelines in Kubernetes Native Mode:**
````
should this be a heading rather than just bold?

gmfrasca commented on README.md:
````diff
@@ -412,6 +412,92 @@ kustomize build applications/pipeline/upstream/env/cert-manager/platform-agnosti
+kubectl get workflows -n kubeflow
+```
+
+**Benefits of Kubernetes Native API Mode:**
````
should this be a heading?

juliusvonkohout commented on README.md:
```diff
@@ -412,6 +412,90 @@ kustomize build applications/pipeline/upstream/env/cert-manager/platform-agnosti
+kustomize build applications/pipeline/upstream/env/cert-manager/platform-agnostic-k8s-native | kubectl apply -f -
+
+# For multi-user deployment
+kustomize build applications/pipeline/upstream/env/cert-manager/platform-agnostic-multi-user-k8s-native | kubectl apply -f -
```
Is there a reason that k8s-native is a separate option? why not incorporate it into the default for KFP ?

VaniHaripriya replied:
k8s-native requires separate deployment option because it requires additional components  like cert-manager, admission webhooks to be deployed which are not used in the default deployment.

juliusvonkohout replied:
but compared to platform-agnostic-multi-user which we already use. Why is it separate ? This is kubeflow platfrom we anyway use cert-manager etc.

juliusvonkohout commented on README.md:
````diff
@@ -412,6 +412,90 @@ kustomize build applications/pipeline/upstream/env/cert-manager/platform-agnosti
+kustomize build applications/pipeline/upstream/env/cert-manager/platform-agnostic-multi-user-k8s-native | kubectl apply -f -
+```
+
+**Key Differences in Kubernetes Native API Mode:**
````
```suggestion
**Main Differences in Kubernetes Native API Mode:**

```

mprahl commented on README.md:
````diff
@@ -412,6 +412,90 @@ kustomize build applications/pipeline/upstream/env/cert-manager/platform-agnosti
 ```
 This installs Argo with the runasnonroot emissary executor. Please note that you are still responsible for analyzing the security issues that arise when containers are run with root access and for deciding if the Kubeflow pipeline main containers are run as runasnonroot. It is generally strongly recommended that all user-accessible OCI containers run with Pod Security Standards [restricted](https ...
 
+#### Kubeflow Pipelines - Kubernetes Native API Mode
````
Could you please restructure this so that it's clear they you can choose between `platform-agnostic-multi-user` and `platform-agnostic-multi-user-k8s-native`?

mprahl commented on README.md:
```diff
@@ -412,6 +412,90 @@ kustomize build applications/pipeline/upstream/env/cert-manager/platform-agnosti
+- Pipeline definitions are stored as `Pipeline` and `PipelineVersion` Custom Resources in Kubernetes
+- No external database required for pipeline storage (ML Metadata still uses external storage)
+- Pipeline validation is handled through Kubernetes admission webhooks
+- Pipeline versioning is managed through Kubernetes resource versioning
```
I think this is a little confusing because resource versioning sounds like you can update the PipelineVersion object but the spec is immutable. Perhaps this line can be removed.

mprahl commented on README.md:
```diff
@@ -412,6 +412,90 @@ kustomize build applications/pipeline/upstream/env/cert-manager/platform-agnosti
+    return hello_task.output
+
+# Compile directly to Kubernetes manifests
+kfp.compiler.Compiler().compile(
```
Could you please add display name in the example?

mprahl commented on README.md:
````diff
@@ -412,6 +412,90 @@ kustomize build applications/pipeline/upstream/env/cert-manager/platform-agnosti
+kubectl get workflows -n kubeflow
+```
+
+#### Benefits of Kubernetes Native API Mode
+
+- **GitOps Integration**: Pipeline definitions can be version controlled and deployed through GitOps workflows
+- **Kubernetes Native**: Leverages Kubernetes RBAC, resource management, and monitoring
+- **Reduced External Dependencies**: Eliminates the need for external database storage for pipeline definitions
+- **Better Security**: Pipeline access control through Kubernetes RBAC
+- **Simplified Operations**: Standard Kubernetes tooling for management and monitoring
+- **Declarative Management**: Treat pipelines as code alongside other Kubernetes resources
+
+**Note**: This mode is ideal for organizations that prefer Kubernetes-native workflows and want to manage pipelines using standard Kubernetes tools and practices.
````
I think we can remove this section since it technically uses the same Kubernetes RBAC as the REST API and the other points are mostly repeating what was said in the key differences section.

VaniHaripriya commented:
@juliusvonkohout @mprahl @gmfrasca Thanks for the reviews! To streamline documentation maintenance, I moved the deployment and compilation instructions to kubeflow/website and added references to them in the manifests repo. I’ve also addressed the comments and updated the PR — please take a look.
Website PR- https://github.com/kubeflow/website/pull/4177

VaniHaripriya commented on README.md:
```diff
@@ -412,6 +412,48 @@ kustomize build applications/pipeline/upstream/env/cert-manager/platform-agnosti
+
+#### Install Kubeflow Pipelines in Kubernetes Native API Mode
+
+For detailed deployment instructions, please refer to the [Kubeflow Pipelines installation guide](https://www.kubeflow.org/docs/components/pipelines/operator-guides/installation/#deploying-kubeflow-pipelines-in-kubernetes-native-api-mode).
```
@juliusvonkohout The reason for seperating them is to let users choose between the standard multi-user setup and the Kubernetes Native API mode, which stores pipelines as CRs and needs cert-manager.

VaniHaripriya replied:
Also, I missed to add platform-agnostic-multi-user-k8s-native uses webhook related [manifests](https://github.com/kubeflow/manifests/tree/master/applications/pipeline/upstream/env/cert-manager/platform-agnostic-multi-user-k8s-native/patches) which are not required in the standard multi-user.

mprahl commented:
/lgtm
